pagefaultgames/pokerogue
# [Bug] Fix a couple small issues with uturn and friends

## What are the changes?
Fixes three main issues with u-turn and company:
- U-turn would trigger regenerator twice on the user
- U-turn would trigger abilities like static twice on the attacker
- The party UI for u-turn and other switch moves would appear in the middle of the move, before effects and abilities were done applying

Note that this _does not_ fix issues with wilds fleeing using uturn, particularly when they flee and activate an ability such as poison point or poison touch at the same time. but it doesn't make it any worse, either - I want to do that in a separate PR


## Why am I doing these changes?
Pre-work for a pursuit implementation




## What did change?
- Removed the manual application of two ability attr types by ForceSwitchOutAttr
- Changed ForceSwitchOutAttr to use SwitchPhase for choosing and sending out a mon, rather than doing it inline in the MoveEffectPhase/apply() function
  - this should drastically improve the stability of the move in edge case scenarios, since the current MoveEffectPhase will complete entirely before the new pokemon is sent out. if additional phases are injected by other effects, though - the phase ordering between moves/abilities may need to be closely analyzed.
- Added UTs for everything that was broken, or that I noticed I had accidentally broke while working on these fixes
  - Added some new helpers/fixes for the UT harnesses to help test switch moves



### Screenshots/Videos


#### Pre-change: U-turn damage and effects (Stamina) happen _after_ party menu
[screen-capture (4).webm](https://github.com/user-attachments/assets/5a9fed96-40c8-4e17-bee3-17b134580326)

#### Pre-change: U-turn triggers regenerator twice
[screen-capture (5).webm](https://github.com/user-attachments/assets/b0569f2f-b487-40f1-af2a-a4f97221eb4a)

#### Pre-change: U-turn triggers static twice (only with no party members to switch to, because... yeah)
[screen-capture (6).webm](https://github.com/user-attachments/assets/79a7a9c7-735b-43db-88dd-ec632e74f664)

#### Pre-change: Demonstration of U-turn in doubles
[screen-capture (9).webm](https://github.com/user-attachments/assets/78ce009d-a077-4f8b-83d4-e4662666ae07)

#### Post-change: U-turn damage and effects (Stamina) happen _before_ party menu
[screen-capture (1).webm](https://github.com/user-attachments/assets/f11fd89c-903d-4a40-acb5-f5ff4c0938d4)

#### Post-change: Baton pass still passes Work Up
[screen-capture (2).webm](https://github.com/user-attachments/assets/d7ac0ea0-988b-430b-9d47-0201f584c0d5)

#### Post-change: Dragon tail activates Stamina on outgoing mon
This still looks a little funky, but it does function correctly

[screen-capture (3).webm](https://github.com/user-attachments/assets/b312a04e-fe57-4f7a-a1a8-da0e3c4adc10)

#### Post-change: U-turn triggers regenerator once
[screen-capture (8).webm](https://github.com/user-attachments/assets/199edadd-ed49-4c1f-9681-be50345ce2cd)

#### Post-change: U-turn triggers static once
[screen-capture (7).webm](https://github.com/user-attachments/assets/2491bf56-1766-4855-b203-ce16d167a7ee)

#### Post-change: Demonstration of U-turn in doubles (works the same except for the better party UI placement)
[screen-capture (10).webm](https://github.com/user-attachments/assets/b63dbdc0-bf91-40bc-a535-757b635459d1)

## How to test the changes?




- automated tests ensure I fixed the regeneration/status ability problem (for status abilities, technically they only ensure the ability still applies, not exactly that it only applies a single time)
- automated tests ensure I didn't break at least the simplest cases for baton pass
- videos can verify the changes to party menu timing

The following overrides are helpful for local testing (make sure you have a party of 2 or 3 depending on if you are using doubles or not):
```ts
const overrides = {
  MOVESET_OVERRIDE: [
    Moves.DRAGON_TAIL,
    Moves.WORK_UP,
    Moves.U_TURN,
    Moves.BATON_PASS
  ],
  STARTING_WAVE_OVERRIDE: 5,
  ABILITY_OVERRIDE: Abilities.REGENERATOR,
  // some useful testing abilities
  OPP_ABILITY_OVERRIDE: Abilities.STAMINA,
  // OPP_ABILITY_OVERRIDE: Abilities.ROUGH_SKIN,
  // OPP_ABILITY_OVERRIDE: Abilities.STATIC,
  OPP_MOVESET_OVERRIDE: new Array(4).fill(Moves.U_TURN),
  OPP_SPECIES_OVERRIDE: Species.SHEDINJA,
  // BATTLE_TYPE_OVERRIDE: 'double',
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;
```

## Checklist
